haskell/src/Main.hs

```diff
@@ -0,0 +1,170 @@
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE MultiWayIf #-}
+module Main where
+
+import UI.NCurses
+import Control.Monad.IO.Class(liftIO)
+import Control.Monad(unless)
+import System.Environment(getArgs)
+import Data.List(sort)
+import Prelude hiding (Either(..))
+
+type Coord = (Int, Int)
+
+data Move = Up | Left | Right | Down
+
+data Level = Level { walls :: [Coord]
+                   , storage :: [Coord]
+                   , crates :: [Coord]
+                   , player :: Coord
+                   , width :: Int
+                   , height :: Int
+                   }
+
+instance Show Level where
+    show = unlines . levelToStrings
+
+levelToStrings :: Level -> [String]
+levelToStrings level@Level{width,height} =
+    [[symbolAt level (x, y) | x <- [0..width]] | y <- [0..height]]
+
+symbolAt :: Level -> Coord -> Char
+symbolAt level coord
```

Comment:
With the aforementioned `Obstacle`, instead of doing a conditional check here, you'd be able to pattern match for most of this too.